### ISSUE TITLE
Raspberry Pi specific improvements

### ISSUE THREAD

**Issue by Botspot** [hakandundar34coding/system-monitoring-center#78]
Raspberry Pi operating systems include a command called `vcgencmd`, which can obtain device-specific information that is not reflected in more global locations. While ideally a system monitor should be able to rely on standardized locations for parameters, that simply is not the case in the ARM world. (Maybe in 5 years...)

[Link to vcgencmd documentation](https://www.raspberrypi.com/documentation/computers/os.html#vcgencmd)

Here are the specific ways in which System-Monitoring-Center can use vcgencmd:
- To determine total RAM on the device, regardless of CPU allocation: `vcgencmd get_config total_mem`
    Example output: `total_mem=8192` (in megabytes)
- To determine memory allocated to the GPU: `vcgencmd get_mem gpu`
    Example output: `gpu=128M` (in megabytes)
- To determine GPU frequency: `vcgencmd get_config gpu_freq`
    Example output: `gpu_freq=700` (in megahertz)

Also I found [a possible way](https://stackoverflow.com/a/63474811) to determine L1/L2/L3 cache size, but it involves compiling a kernel module. :( This is probably more trouble than it's worth, but is still interesting to document.

**Comment by hakandundar34coding** [comment]
Can you share output of these python commands in order to learn the time for running them?

```
import time, subprocess

time_start = time.time()
output = (subprocess.check_output(["vcgencmd", "get_config", "total_mem"], shell=False)).decode()
print(time.time() - time_start)

time_start = time.time()
output = (subprocess.check_output(["vcgencmd", "get_mem", "gpu"], shell=False)).decode()
print(time.time() - time_start)
```

What is the difference between these commands? There is no information about `get_config` in this [page](https://pypi.org/project/vcgencmd/).
`vcgencmd measure_clock v3d` and `vcgencmd get_config gpu_freq`

Does output of `vcgencmd get_config gpu_freq` changes if there is more/less load on GPU?

**Comment by Botspot** [comment]
> Can you share output of these python commands in order to learn the time for running them?

Sure, but I'll be using the handy `time` command instead of your python stuff.
```
$ time vcgencmd get_config total_mem
total_mem=8192

real	0m0.004s
user	0m0.001s
sys	0m0.004s
```
```
$ time vcgencmd get_mem gpu
gpu=128M

real	0m0.006s
user	0m0.001s
sys	0m0.003s
```

> What is the difference between these commands? There is no information about `get_config` in this [page](https://pypi.org/project/vcgencmd/). `vcgencmd measure_clock v3d` and `vcgencmd get_config gpu_freq`

Interesting. The `measure_clock` option was not even mentioned in the official documentation. Yes, it does appear to change based on usage, unlike `get_config` which stays constant and may not reflect current values.
I would recommend getting the frequency of `core`, rather than `v3d`, and this is the output on my Pi4:
```
pi@raspberrypi:~ $ vcgencmd measure_clock core
frequency(1)=699996096
```

**Comment by hakandundar34coding** [comment]
The commands are run very fast. They can be run repeatedly even in less than a second.

I will add support for this tool (vcgencmd). It will be usable in v1.17.0 (probably in a week). Because package for v1.16.0 is prepared, tested and it will be downloadable tomorrow.

**Comment by hakandundar34coding** [comment]
I have updated the source [code](https://github.com/hakandundar34coding/system-monitoring-center/archive/refs/heads/master.zip).

`vcgencmd` dependency will be optional. It will not be installed during installation of this application.

Support for GPU frequency and memory of Broadcom GPUs is added. 

Support for physical RAM size of RB Pi devices is not added. Because `vcgencmd get_config total_mem` command uses `get_config` and the value may be changed by editing some configuration files. I do not know. Do you have any information about this situation?

Can you run the application from the source code and share commandline output (if there are errors) and screenshot of the GPU tab?

For running the application from the source code:
You can use `python3 Main.py` command in the `src` folder.

**Comment by Botspot** [comment]
> Support for physical RAM size of RB Pi devices is not added. Because `vcgencmd get_config total_mem` command uses `get_config` and the value may be changed by editing some configuration files. I do not know. Do you have any information about this situation?

There is no reason to have reservations for including `vcgencmd get_config total_mem`. This is the only way to detect the full RAM chip's size on a Raspberry Pi. Alternative methods will only see the amount of RAM dedicated to the CPU.
Concerning the possibility of the user changing configuration on-the-fly, don't worry about it. Raspberry Pi systems have a low-level config file called `config.txt` in the boot partition. This file can be modified during runtime, but the changes won't take effect until the next reboot.
Besides, while it is technically possible to limit usable ram in /boot/config.txt, who would want to do that?! 🤣

**Comment by hakandundar34coding** [comment]
Physical RAM size support is also added.

Editing this value by user is a very very limited case. I did not want to use this value. Currently it is used by the code. It may be removed if there are problems about the configurations.

**Comment by Botspot** [comment]
> Editing this value by user is a very very limited case. I did not want use this value.

Under no circumstances would System Monitoring Center display incorrect information. For the user to change available RAM, they would have to edit /boot/config.txt. But those changes will not take effect until next reboot. So until the next reboot, System Monitoring Center would still display the *current* RAM size correctly.
After a reboot, the new configuration would take effect and only then would `vcgencmd get_config total_mem` reflect the new value.
Basically what I'm trying to say is that this get_config option is 100% correct in all scenarios.

I will test your source code in a few minutes.

**Comment by Botspot** [comment]
> I will test your source code in a few minutes.

Everything looks good on the GPU tab.
The RAM tab still incorrectly shows the amount available to the CPU, not the total amount. (7.2GB instead of 8.0GB)

During testing, I noticed a bug in your CPU tab. Try selecting a different core than `cpu0`, and notice that "Min-Max frequency" vanishes, and "Average Usage" freezes.

**Comment by hakandundar34coding** [comment]
About the RAM tab:
Physical RAM size is shown above the graph. RAM capacity (usable) is shown below the graph.
Which one is 7.2 GB?

Here is an example screenshot:
![smc_ram](https://user-images.githubusercontent.com/90193017/169588645-0f10fe34-ace0-4834-bea8-56eccd231cec.png)


About the CPU tab:
It works on my system (x86_64). Is there any error when the application is run from the terminal?

**Comment by Botspot** [comment]
> About the CPU tab:
> It works on my system (x86_64). Is there any error when the application is run from the terminal?

There are no errors when it first launches. The CPU tab starts on cpu0 by default.
But when I select cpu1, this output appears, repeating once per second:
```
Traceback (most recent call last):
  File "/home/pi/system-monitoring-center/src/MainGUI.py", line 575, in on_row_activated
    Cpu.cpu_loop_func()
  File "/home/pi/system-monitoring-center/src/Cpu.py", line 160, in cpu_loop_func
    cpu_core_current_frequency = self.cpu_core_current_frequency_func(selected_cpu_core_number_only)
  File "/home/pi/system-monitoring-center/src/Cpu.py", line 368, in cpu_core_current_frequency_func
    return cpu_core_current_frequency
UnboundLocalError: local variable 'cpu_core_current_frequency' referenced before assignment
```

**Comment by hakandundar34coding** [comment]
Is there a file like that?
`/sys/devices/system/cpu/cpufreq/policy1/scaling_cur_freq`

**Comment by Botspot** [comment]
This file exists: /sys/devices/system/cpu/cpufreq/policy0/scaling_cur_freq
This file does not: /sys/devices/system/cpu/cpufreq/policy1/scaling_cur_freq

**Comment by hakandundar34coding** [comment]
I will update the source code for the frequency bug tomorrow.
It looks like frequency information is not provided for all CPU cores on RB-Pi devices.
Can you share output of this command?
`ls /sys/devices/system/cpu/cpufreq`

What about the RAM [information](https://github.com/hakandundar34coding/system-monitoring-center/issues/78#issuecomment-1133187258)?

**Comment by Botspot** [comment]
> ls /sys/devices/system/cpu/cpufreq

I'll use `tree` so you can see the contents of each folder.
```
$ tree /sys/devices/system/cpu/cpufreq
/sys/devices/system/cpu/cpufreq
├── ondemand
│   ├── ignore_nice_load
│   ├── io_is_busy
│   ├── powersave_bias
│   ├── sampling_down_factor
│   ├── sampling_rate
│   └── up_threshold
└── policy0
    ├── affected_cpus
    ├── cpuinfo_cur_freq
    ├── cpuinfo_max_freq
    ├── cpuinfo_min_freq
    ├── cpuinfo_transition_latency
    ├── related_cpus
    ├── scaling_available_frequencies
    ├── scaling_available_governors
    ├── scaling_cur_freq
    ├── scaling_driver
    ├── scaling_governor
    ├── scaling_max_freq
    ├── scaling_min_freq
    ├── scaling_setspeed
    └── stats
        ├── reset
        ├── time_in_state
        ├── total_trans
        └── trans_table
```

> What about the RAM [information](https://github.com/hakandundar34coding/system-monitoring-center/issues/78#issuecomment-1133187258)?

Both locations are incorrect.
![2022-05-20-161154_1920x1080_scrot](https://user-images.githubusercontent.com/54716352/169611220-e565df61-3ea9-4315-89c9-51747e610a0e.png)

**Comment by hakandundar34coding** [comment]
About the RAM tab:
The second value is correct. The source code is updated again for the first one.

About the CPU tab:
Probably, CPU current frequencies are same for all cores on RPi devices. There is a similar behavior for the frequencies of the 
several components of the GPUs of the RPis. 

Content of the `/sys/devices/system/cpu/cpufreq/policy0/related_cpus` (or `.../affected_cpus`) file may be useful for understanding this behavior.

Value for only the first core is provided in the  `/sys/devices/system/cpu/cpufreq/policy0/scaling_cur_freq` file.

The source code is updated and frequency of the first core is used for the other cores if a separate file is not found.

**Comment by Botspot** [comment]
I've issued `git pull` and not much changed. The RAM tab still shows 7.6GB in both places, and the "Min-Max frequency" still vanishes for any core other than cpu0. There were no useful console errors this time.
I did confirm, however, that "Average Usage" continued working when viewing cpu1, cpu2, and cpu3.

Here are the files that you requested:
```
$ cat /sys/devices/system/cpu/cpufreq/policy0/related_cpus
0 1 2 3
```
```
$ cat /sys/devices/system/cpu/cpufreq/policy0/affected_cpus
0 1 2 3
```

**Comment by hakandundar34coding** [comment]
The source code is updated again.

Only frequency of first core is available on your system. All core frequencies are same. But a more complex algorithm is is required to get frequencies of different cores if there are big and little cores on different systems (CPU cores on your system may be same). 

In order to avoid this difficulty:
A different file is used for getting core frequencies. The newer file is a link to the previous one. But this time there is no need to use more code for ARM CPU cores. Can you write it if current and min-max frequencies are shown as numeric values when different CPU cores are selected?

There are cpufreq folders (links) under folders of all cores. Probably there are same folders for RPis.
```
/sys/devices/system/cpu/cpu0/cpufreq
/sys/devices/system/cpu/cpu1/cpufreq
/sys/devices/system/cpu/cpu2/cpufreq
.
.
```

About the RAM tab:
There is no change. But some lines are added for printing some numbers/values to learn where is the problem. Can you wite terminal output?

**Comment by Botspot** [comment]
> In order to avoid this difficulty: A different file is used for getting core frequencies. The newer file is a link to the previous one. But this time there is no need to use more code for ARM CPU cores.
> 
> There are cpufreq folders (links) under folders of all cores. Probably there are same folders for RPis.
> 
> ```
> /sys/devices/system/cpu/cpu0/cpufreq
> /sys/devices/system/cpu/cpu1/cpufreq
> /sys/devices/system/cpu/cpu2/cpufreq
> .
> .
> ```

This appears to be workable. I checked the files on my system:
```
$ file /sys/devices/system/cpu/cpu0/cpufreq
/sys/devices/system/cpu/cpu0/cpufreq: symbolic link to ../cpufreq/policy0
$file /sys/devices/system/cpu/cpu1/cpufreq
/sys/devices/system/cpu/cpu1/cpufreq: symbolic link to ../cpufreq/policy0
```

> Can you write if it current and min-max frequencies are shown as numeric values when different CPU cores are selected?

Min-Max frequencies are working for all cores now.

> About the RAM tab: There is no change. But some lines are added for printing some numbers/values to learn where is the problem. Can you wite terminal output?

When I clicked the Memory tab, these two lines appeared in the terminal.
```
1
4
```
Other than that, no output was produced; not even GTK warnings anymore.

**Comment by theofficialgman** [comment]
@Botspot what is the actual output from the command: `vcgencmd get_config total_mem` ?

its pretty normal for the capacity to be lower than 8GiB (Gibibytes is the actual unit) due to memory carveouts that prevent normal access (like carevouts for kernel, display buffers, gpu, etc)

for reference, on 1.16.0, my capacity (at the bottom) on switch is slightly lower than the physical ram value (at the top)

**Comment by Botspot** [comment]
> @Botspot what is the actual output from the command: `vcgencmd get_config total_mem` ?

```
pi@raspberrypi:~ $ vcgencmd get_config total_mem
total_mem=8192
```

**Comment by hakandundar34coding** [comment]
There is a problem when running `vcgencmd get_config total_mem` command by using Python subprocess module..

Can you write these lines into `memory_test.py` file and run it by using Python and write output?

```
import subprocess
total_physical_ram = (subprocess.check_output(["vcgencmd", "get_config", "total_mem"], shell=False)).decode()
```

But it works when `gpu_memory_capacity = (subprocess.check_output(["vcgencmd", "get_mem", "gpu"], shell=False)).decode()` is used for GPU tab.

**Comment by Botspot** [comment]
The problem could be due to `vcgencmd get_mem gpu` returning a value ending with `M`, but `get_config total_mem` does not return any measurement unit with it. That's the only difference I see.

**Comment by Botspot** [comment]
> Can you write these lines into `memory_test.py` file and run it by using Python and write output?

I created a new file called `memory_test.py` and placed your two lines in it.
When running it, there was no output.
```
$ python3 ~/memory_test.py
$ 
```

**Comment by hakandundar34coding** [comment]
M is not a problem. The value is converted into a float value directly.

Finally,
What happens if you append these lines into memory_test.py file and run it again?
`print(total_physical_ram)`
`print(float(total_physical_ram.strip().split("=")[1]))`

**Comment by Botspot** [comment]
Full output:
```
$ python3 ~/memory_test.py
total_mem=8192

8192.0
```

**Comment by hakandundar34coding** [comment]
I disabled the error handling codes.

You can try running the latest [code](https://github.com/hakandundar34coding/system-monitoring-center/archive/refs/heads/master.zip) to see if there are problems if you want to test it.

**Comment by Botspot** [comment]
> You can try running the latest code to see if there are problems if you want to test it.

Now it's blank.
![Screenshot from 2022-05-23 10-27-49](https://user-images.githubusercontent.com/54716352/169854233-b33bbdb8-76e6-4814-9774-7c7dfa738020.png)

**Comment by hakandundar34coding** [comment]
What about the errors? :)

**Comment by Botspot** [comment]
```
Traceback (most recent call last):
  File "/home/pi/system-monitoring-center/src/Memory.py", line 159, in memory_initial_func
    total_physical_ram = (subprocess.check_output(["vcgencmd", "get_config", "total_mem"], shell=False)).decode().strip().split("=")[1]
NameError: name 'subprocess' is not defined
```

**Comment by hakandundar34coding** [comment]
The problem is fixed now. Can you test the latest source code again and share screenshot/terminal output?

**Comment by Botspot** [comment]
It works now! No errors, and the upper Memory monitor shows 8.0 GiB as expected.

Before I close this issue, I noticed one other bonus problem. Too many /dev/ block devices show up in the "Disk" tab.
![2022-05-23-103518_1920x1080_scrot](https://user-images.githubusercontent.com/54716352/169855684-a479e260-31d2-4029-8189-552a5ee309e5.png)
In addition to ram0..ram15, it is also showing loop0, loop1, zram0, and zram1.
I would recommend filtering the drives' "Major numbers" to only show valid writable media. In the bash world, you would do this: `lsblk -I 8,179,259 -rno NAME`

**Comment by hakandundar34coding** [comment]
There was an [issue](https://github.com/hakandundar34coding/system-monitoring-center/issues/81) for an option for hiding loop devices.

I will close this issue. Comments for hiding loop devices and ram[NUMBER] disks can be written in the relevant issue.

I will reopen this issue if you write additional bugs for RPi devices.
